File: hmh-frontend/src/api/stock.ts
import client from "./client";

export type MovementType =
  | "OPENING_BALANCE" | "DELIVERY_RECEIVED" | "USAGE"
  | "ADJUSTMENT_ADD" | "ADJUSTMENT_SUBTRACT"
  | "RETURN_TO_STORE" | "TRANSFER_IN" | "TRANSFER_OUT";

export interface StockLedgerEntry {
  id: string;
  project_id: string;
  site_id: string;
  lot_id: string | null;
  item_id: string;
  boq_item_id: string | null;
  movement_type: MovementType;
  reference_type: string;
  reference_id: string | null;
  quantity_in: number;
  quantity_out: number;
  unit: string | null;
  unit_cost: number | null;
  movement_date: string;
  entered_by: string | null;
  notes: string | null;
  created_at: string;
}

export interface StockBalance {
  project_id: string;
  site_id: string;
  lot_id: string | null;
  item_id: string;
  balance: number;
  last_movement_date: string | null;
  item_name: string | null;
  item_unit: string | null;
}

export interface UsageLog {
  id: string;
  project_id: string;
  site_id: string;
  lot_id: string | null;
  stage_id: string | null;
  item_id: string;
  boq_item_id: string | null;
  quantity_used: number;
  used_by_person_name: string | null;
  used_by_team_name: string | null;
  recorded_by_user_id: string;
  usage_date: string;
  comments: string | null;
  created_at: string;
}

export interface UsageLogCreate {
  site_id: string;
  item_id: string;
  quantity_used: number;
  lot_id?: string | null;
  stage_id?: string | null;
  boq_item_id?: string | null;
  used_by_person_name?: string | null;
  used_by_team_name?: string | null;
  usage_date?: string | null;
  comments?: string | null;
}

export const stockApi = {
  getBalances: async (projectId: string, siteId?: string): Promise<StockBalance[]> => {
    const res = await client.get<{ data: StockBalance[] }>("/stock/balances", {
      params: { project_id: projectId, ...(siteId ? { site_id: siteId } : {}) },
    });
    return res.data.data;
  },

  getLedger: async (
    projectId: string,
    params?: { site_id?: string; item_id?: string; limit?: number }
  ): Promise<StockLedgerEntry[]> => {
    const res = await client.get<{ data: StockLedgerEntry[] }>("/stock/ledger", {
      params: { project_id: projectId, ...params },
    });
    return res.data.data;
  },

  recordUsage: async (projectId: string, body: UsageLogCreate): Promise<UsageLog> => {
    const res = await client.post<{ data: UsageLog }>("/stock/usage", body, {
      params: { project_id: projectId },
    });
    return res.data.data;
  },
};

File: hmh-backend/app/schemas/stock.py
"""Pydantic v2 schemas for stock ledger, balances, and usage."""

import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict

from app.models.enums import MovementType


class StockLedgerRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    site_id: uuid.UUID
    lot_id: Optional[uuid.UUID] = None
    item_id: uuid.UUID
    boq_item_id: Optional[uuid.UUID] = None
    movement_type: MovementType
    reference_type: str
    reference_id: Optional[uuid.UUID] = None
    quantity_in: float
    quantity_out: float
    unit: Optional[str] = None
    unit_cost: Optional[float] = None
    movement_date: datetime
    entered_by: Optional[uuid.UUID] = None
    notes: Optional[str] = None
    created_at: datetime


class StockBalanceRead(BaseModel):
    """Represents a row from the stock_balances materialized view."""
    project_id: uuid.UUID
    site_id: uuid.UUID
    lot_id: Optional[uuid.UUID] = None
    item_id: uuid.UUID
    balance: float
    last_movement_date: Optional[datetime] = None

    # Populated by joining items table
    item_name: Optional[str] = None
    item_unit: Optional[str] = None


class UsageLogRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    project_id: uuid.UUID
    site_id: uuid.UUID
    lot_id: Optional[uuid.UUID] = None
    stage_id: Optional[uuid.UUID] = None
    item_id: uuid.UUID
    boq_item_id: Optional[uuid.UUID] = None
    quantity_used: float
    used_by_person_name: Optional[str] = None
    used_by_team_name: Optional[str] = None
    recorded_by_user_id: uuid.UUID
    usage_date: datetime
    comments: Optional[str] = None
    created_at: datetime


class UsageLogCreate(BaseModel):
    site_id: uuid.UUID
    item_id: uuid.UUID
    quantity_used: float
    lot_id: Optional[uuid.UUID] = None
    stage_id: Optional[uuid.UUID] = None
    boq_item_id: Optional[uuid.UUID] = None
    used_by_person_name: Optional[str] = None
    used_by_team_name: Optional[str] = None
    usage_date: Optional[datetime] = None
    comments: Optional[str] = None
